File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.py
import sys
sys.setrecursionlimit(10 ** 6)

def solution(numbers, target):
    answer = 0
    sumNum = -numbers[0] 
    
    def DFS(deep, sumNum):
        nonlocal answer
        if(deep == len(numbers)):
            if(sumNum == target):
                answer += 1
            return
        DFS(deep + 1, sumNum + numbers[deep])
        DFS(deep + 1, sumNum - numbers[deep])
    
    DFS(1, sumNum)
    sumNum = numbers[0]
    DFS(1, sumNum)
    return answer

File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    private static int sum, count = 0;
    
    private static void dfs(int deep, int[] numbers, int target, int sum) {
        if(deep == numbers.length) {
            if(sum == target) {
                count += 1;
            }
            return;
        }
        for(int i = 0 ; i < 2 ; i++) {
            if(i == 0) { // +
                dfs(deep + 1, numbers, target, sum + numbers[deep]);
            }
            else if(i == 1) { // -
                dfs(deep + 1, numbers, target, sum - numbers[deep]);
            }
        }
    }
    
    public int solution(int[] numbers, int target) {
        
        dfs(0, numbers, target, sum);
        
        return count;
    }
}